ecs-agent/acs/session/task_manifest_responder.go

```diff
@@ -0,0 +1,177 @@
+		taskComparer:      taskComparer,
+		snAccessor:        snAccessor,
+		messageIDAccessor: messageIDAccessor,
+		cluster:           cluster,
+		logger:            logrus.WithField("ACSRequestResponder", "TaskManifestResponder"),
+		metricsFactory:    metricsFactory,
+	}
+	r.respond = ResponseToACSSender(r.Name(), responseSender)
+	return r
+}
+
+func (*taskManifestResponder) Name() string { return "task manifest responder" }
+
+func (tmr *taskManifestResponder) HandlerFunc() wsclient.RequestHandler {
+	return tmr.handleTaskManifestMessage
+}
+
+// handleTaskManifestMessage is the high level caller to handle a task manifest message from ACS.
+// It will kick off the call stack of processTaskManifestMessage calling ack and send TaskManifestMessage
+func (tmr *taskManifestResponder) handleTaskManifestMessage(message *ecsacs.TaskManifestMessage) error {
+	tmr.logger.WithField("MessageId", aws.StringValue(message.MessageId)).Info("Processing task manifest message")
+	err := tmr.processTaskManifestMessage(message)
+	tmr.metricsFactory.New(metrics.TaskManifestHandlingDuration).Done(err)
+	return err
+}
+
+// processTaskManifestMessage processes a task manifest message from ACS. It verifies
+// whether a task manifest message is stale or not. If not, it proceeds to create a
+// list of stop candidates.
+func (tmr *taskManifestResponder) processTaskManifestMessage(
+	message *ecsacs.TaskManifestMessage) error {
+	manifestSeqNum := aws.Int64Value(message.Timeline)
+	agentLatestSeqNum := tmr.snAccessor.GetLatestSequenceNumber()
+
+	// Verify that task manifest isn't stale.
+	// The agent will keep track of the highest sequence number received in both PayloadMessage and TaskManifestMessage.
+	// If the sequence number on TaskManifestMessage is lower to the latest known one, the TaskManifestMessage must be discarded.
+	// ACS will guarantee that the sequence number is indeed increasing and valid for this use case.
+	if manifestSeqNum < agentLatestSeqNum {
+		tmr.logger.WithFields(logrus.Fields{
+			"ManifestSeqNum":    manifestSeqNum,
+			"AgentLatestSeqNum": agentLatestSeqNum,
+		}).Warn("Discarding task manifest message. " +
+			"Sequence number from manifest is less than Agent latest sequence number.")
+		return nil
+	}
+
+	// Save the task manifest's messageID and sequence number so that it can be compared
+	// by the taskStopVerificationAckResponder.
+	err := tmr.messageIDAccessor.SetMessageID(aws.StringValue(message.MessageId))
+	if err != nil {
+		return errors.Wrap(err, "failed to update latest manifest messageID")
+	}
+	err = tmr.snAccessor.SetLatestSequenceNumber(message)
+	if err != nil {
+		return errors.Wrap(err, "failed to update latest manifest sequence number")
+	}
+
+	// Get list of tasks on instance.
+	tasksToMessage, err := tmr.taskComparer.CompareRunningTasksOnInstanceWithManifest(tmr, message)
+	if err != nil {
+		return err
+	}
+
+	// Acknowledge the task manifest message asynchronously.
+	var wg sync.WaitGroup
```

Comment:
I am curious what we gain by this asynchronicity. IIUC there is only a single TCP connection between Agent and ACS at any time, so the two messages will be sent synchronously anyway?

Reply:
Discussed with @timj-hh offline. The current behavior of Agent is to process the sending of ack and task stop verification messages asynchronously and this PR is only for refactoring and not to change any behavior. So, we would like to keep the asynchronicity here.

Reply:
Reference - https://github.com/aws/amazon-ecs-agent/blob/e978160b886ddfce9120a21beac9e09eca1bda92/agent/acs/handler/task_manifest_handler.go#L92-L93